[flaky-test] TestDBResiliencyYugabyteScenarios/tablet

```
      db_node_failure_handling_test.go:62: starting yuga cluster with (3) masters and (3) tablets 
      db_node_failure_handling_test.go:62: network sc_yuga_net_9d563c00-ca67-434e-b024-019ca99ff425 created
      yugabyte.go:149: executing /home/yugabyte/bin/yb-admin -master_addresses yuga-master-11a729b3-28ed-46fb-ade9-58e1e913f23c:7100,yuga-master-fa33b571-1a3b-4ef4-bc10-cab69de68507:7100,yuga-master-a350d4ab-f596-4c50-b380-efa0963e9197:7100 list_all_masters
      db_node_failure_handling_test.go:63: Making DB env
  creating configuration files ... ok
      db_node_failure_handling_test.go:63: connection nodes details: 172.20.0.5:5433,172.20.0.6:5433,172.20.0.7:5433
      db_node_failure_handling_test.go:63: [TestDBResiliencyYugabyteScenarios/tablet] db name: sc_test_l45mluyuujqfnnfstvzzo6n4e6ot2n6hgjgexcdpp2gevxedo6mq
  2025/11/18 12:54:31 connection count for 172.20.0.5 going negative!
  2025-11-18T12:54:31.192Z	INFO	dbtest/connection.go:131	[172.19.0.5:5433,172.19.0.6:5433,172.19.0.7:5433] Connected to database
  2025/11/18 12:54:31 connection count for 172.19.0.5 going negative!
      db_node_failure_handling_test.go:63: Allocating ports
  2025-11-18T12:54:32.638Z	INFO	connection/server_util.go:107	Listening on: tcp://localhost:32893
  2025-11-18T12:54:32.638Z	INFO	connection/server_util.go:107	Listening on: tcp://localhost:45453
  2025-11-18T12:54:32.639Z	INFO	connection/server_util.go:107	Listening on: tcp://localhost:38135
  2025-11-18T12:54:32.639Z	INFO	connection/server_util.go:107	Listening on: tcp://localhost:33091
  2025-11-18T12:54:32.639Z	INFO	connection/server_util.go:107	Listening on: tcp://localhost:39355
  2025-11-18T12:54:32.640Z	INFO	connection/server_util.go:107	Listening on: tcp://localhost:34197
  2025-11-18T12:54:32.640Z	INFO	connection/server_util.go:107	Listening on: tcp://localhost:45937
  2025-11-18T12:54:32.640Z	INFO	connection/server_util.go:107	Listening on: tcp://localhost:38895
  2025-11-18T12:54:32.641Z	INFO	connection/server_util.go:107	Listening on: tcp://localhost:42401
  2025-11-18T12:54:32.641Z	INFO	connection/server_util.go:107	Listening on: tcp://localhost:32961
  2025-11-18T12:54:32.641Z	INFO	connection/server_util.go:107	Listening on: tcp://localhost:39725
  2025-11-18T12:54:32.641Z	INFO	connection/server_util.go:107	Listening on: tcp://localhost:41347
  2025-11-18T12:54:32.642Z	INFO	connection/server_util.go:107	Listening on: tcp://localhost:33993
  make: *** [Makefile:110: test-integration-db-resiliency] Error 1
  2025-11-18T12:54:32.642Z	INFO	connection/server_util.go:107	Listening on: tcp://localhost:35585
  2025-11-18T12:54:32.642Z	INFO	connection/server_util.go:107	Listening on: tcp://localhost:44723
  2025-11-18T12:54:32.643Z	INFO	connection/server_util.go:107	Listening on: tcp://localhost:42525
  2025-11-18T12:54:32.643Z	INFO	connection/server_util.go:107	Listening on: tcp://localhost:35671
  2025-11-18T12:54:32.643Z	INFO	connection/server_util.go:107	Listening on: tcp://localhost:46047
      db_node_failure_handling_test.go:63: System Parameters:
          BlockSize: 500
          BlockTimeout: 2e+09
          ClientTLS.CACertPaths: <nil>
          ClientTLS.CertPath: 
          ClientTLS.KeyPath: 
          ClientTLS.Mode: 
          ConfigBlockPath: /tmp/TestDBResiliencyYugabyteScenariostablet1449594508/002/config-block.pb.bin
          DB.Endpoints.0: 172.20.0.5:5433
          DB.Endpoints.1: 172.20.0.6:5433
          DB.Endpoints.2: 172.20.0.7:5433
          DB.LoadBalance: true
          DB.Name: sc_test_l45mluyuujqfnnfstvzzo6n4e6ot2n6hgjgexcdpp2gevxedo6mq
          DB.Password: yugabyte
          DB.TLS.CACertPath: 
          DB.TLS.Mode: 
          Endpoints.Coordinator.Metrics: localhost:35585
          Endpoints.Coordinator.Server: localhost:33993
          Endpoints.LoadGen.Metrics: localhost:46047
          Endpoints.LoadGen.Server: localhost:35671
          Endpoints.Orderer.0.Metrics: localhost:45453
          Endpoints.Orderer.0.Server: localhost:32893
          Endpoints.Query.Metrics: localhost:41347
          Endpoints.Query.Server: localhost:39725
          Endpoints.Sidecar.Metrics: localhost:42525
          Endpoints.Sidecar.Server: localhost:44723
          Endpoints.VCService.0.Metrics: localhost:38895
          Endpoints.VCService.0.Server: localhost:45937
          Endpoints.VCService.1.Metrics: localhost:32961
          Endpoints.VCService.1.Server: localhost:42401
          Endpoints.Verifier.0.Metrics: localhost:33091
          Endpoints.Verifier.0.Server: localhost:38135
          Endpoints.Verifier.1.Metrics: localhost:34197
          Endpoints.Verifier.1.Server: localhost:39355
          LedgerPath: /tmp/TestDBResiliencyYugabyteScenariostablet1449594508/001
          LoadGenBlockLimit: 0
          LoadGenTXLimit: 0
          LoadGenWorkers: 1
          Logging.Caller: true
          Logging.Development: true
          Logging.Enabled: true
          Logging.Level: INFO
          Logging.Name: 
          Logging.Output: 
          Policy.ChannelID: channel1
          Policy.ConfigBlockPath: 
          Policy.Identity: <nil>
          Policy.NamespacePolicies.0.KeyPath: <nil>
          Policy.NamespacePolicies.0.Scheme: ECDSA
          Policy.NamespacePolicies.0.Seed: 3.852159813000522e+18
          Policy.NamespacePolicies.1.KeyPath: <nil>
          Policy.NamespacePolicies.1.Scheme: ECDSA
          Policy.NamespacePolicies.1.Seed: 8.532807521486154e+18
          Policy.NamespacePolicies.2.KeyPath: <nil>
          Policy.NamespacePolicies.2.Scheme: ECDSA
          Policy.NamespacePolicies.2.Seed: 3.8883023510454907e+18
          Policy.NamespacePolicies.3.KeyPath: <nil>
          Policy.NamespacePolicies.3.Scheme: ECDSA
          Policy.NamespacePolicies.3.Seed: 4.512466281294657e+18
          Policy.NamespacePolicies._meta.KeyPath: <nil>
          Policy.NamespacePolicies._meta.Scheme: ECDSA
          Policy.NamespacePolicies._meta.Seed: 5.221277731205827e+18
          Policy.OrdererEndpoints.0.host: localhost
          	Test:       	TestDBResiliencyYugabyteScenarios/tablet
  2025-11-18T12:56:34.059Z	INFO	connection/client_util.go:149	Closing 5 connections.
  [e]orderer] 2025-11-18T12:56:34.059Z	INFO	orderer	mock/orderer.go:159	Finished broadcast with 127.0.0.1:60734
  2025-11-18T12:56:34.059Z	ERROR	connection/client_util.go:161	failed closing connections: rpc error: code = Canceled desc = grpc: the client connection is closing
  rpc error: code = Canceled desc = grpc: the client connection is closing
  rpc error: code = Canceled desc = grpc: the client connection is closing
  rpc error: code = Canceled desc = grpc: the client connection is closing
  2025-11-18T12:56:34.059Z	INFO	connection/client_util.go:149	Closing 5 connections.
  2025-11-18T12:56:34.059Z	ERROR	connection/client_util.go:161	failed closing connections: rpc error: code = Canceled desc = grpc: the client connection is closing
  rpc error: code = Canceled desc = grpc: the client connection is closing
  rpc error: code = Canceled desc = grpc: the client connection is closing
  rpc error: code = Canceled desc = grpc: the client connection is closing
  2025-11-18T12:56:34.078Z	INFO	vc/database.go:99	closing database connection
  2025/11/18 12:56:34 connection count for 172.20.0.5 going negative!
      yugabyte.go:112: stopping and removing node: yuga-master-11a729b3-28ed-46fb-ade9-58e1e913f23c
  2025/11/18 12:56:35 connection count for 172.20.0.5 going negative!
  *** Aborted at 1763470595 (unix time) try "date -d @1763470595" if you are using GNU date ***
  PC: @                0x0 (unknown)
  *** SIGTERM (@0x1) received by PID 7 (TID 0x7f2866cb2340) from PID 1; stack trace: ***
      @     0x7f2865d07371 __pthread_cond_wait
      @     0x555d3762498b yb::TerminationMonitor::WaitForTermination()
      @     0x555d371e6b12 main
      @     0x7f2865f537e5 __libc_start_main
      @     0x555d36fec02e _start
      yugabyte.go:112: Container ada04c0fb27ac3a52587d0f5629e8918d2cb17098bac9e817a7b0445f3b047e6 stopped and removed successfully
      yugabyte.go:112: stopping and removing node: yuga-master-fa33b571-1a3b-4ef4-bc10-cab69de68507
  *** Aborted at 1763470595 (unix time) try "date -d @1763470595" if you are using GNU date ***
  PC: @                0x0 (unknown)
  *** SIGTERM (@0x1) received by PID 7 (TID 0x7f35bdf2d340) from PID 1; stack trace: ***
      @     0x7f35bcf82371 __pthread_cond_wait
      @     0x55f25448298b yb::TerminationMonitor::WaitForTermination()
      @     0x55f254044b12 main
      @     0x7f35bd1ce7e5 __libc_start_main
      @     0x55f253e4a02e _start
      yugabyte.go:112: Container 7f74eac3f5681238e81aa32aadc4311ecb7b4cfab9d7838552936a16a560e515 stopped and removed successfully
      yugabyte.go:112: stopping and removing node: yuga-master-a350d4ab-f596-4c50-b380-efa0963e9197
  *** Aborted at 1763470595 (unix time) try "date -d @1763470595" if you are using GNU date ***
  PC: @                0x0 (unknown)
  *** SIGTERM (@0x1) received by PID 7 (TID 0x7f78d9fcb340) from PID 1; stack trace: ***
      @     0x7f78d9020371 __pthread_cond_wait
      @     0x55de4957098b yb::TerminationMonitor::WaitForTermination()
      @     0x55de49132b12 main
      @     0x7f78d926c7e5 __libc_start_main
      @     0x55de48f3802e _start
      yugabyte.go:112: Container 38c7538d8051cf3717f38b275d9d5a851f00711932a373f9ea19388609cbb639 stopped and removed successfully
      yugabyte.go:112: stopping and removing node: yuga-tablet-abe54694-70a1-4b3d-941c-dac444f695c4
  *** Aborted at 1763470595 (unix time) try "date -d @1763470595" if you are using GNU date ***
  PC: @                0x0 (unknown)
  *** SIGTERM (@0x1) received by PID 7 (TID 0x7f1c5b722340) from PID 1; stack trace: ***
      @     0x7f1c5a777371 __pthread_cond_wait
      @     0x564dda17398b yb::TerminationMonitor::WaitForTermination()
      @     0x564ddb7e8969 yb::tserver::TabletServerMain()
      @     0x564dd9d31ce4 main
      @     0x7f1c5a9c37e5 __libc_start_main
      @     0x564dd9b3b02e _start
      yugabyte.go:112: Container 827a99189f0ef792f1827a831c8bc258cacccd3b2a7f56bbb836e6a4344816c5 stopped and removed successfully
      yugabyte.go:112: stopping and removing node: yuga-tablet-d4ea501f-a14d-4e7b-a791-1dc70224a660
  *** Aborted at 1763470595 (unix time) try "date -d @1763470595" if you are using GNU date ***
  PC: @                0x0 (unknown)
  *** SIGTERM (@0x1) received by PID 7 (TID 0x7f860846f340) from PID 1; stack trace: ***
      @     0x7f86074c4371 __pthread_cond_wait
      @     0x560492fd298b yb::TerminationMonitor::WaitForTermination()
      @     0x560494647969 yb::tserver::TabletServerMain()
      @     0x560492b90ce4 main
      @     0x7f86077107e5 __libc_start_main
      @     0x56049299a02e _start
      yugabyte.go:112: Container 35a2ef39810e4fb33587745397fc1c18f6fc0943f8429cf4563cec5c8de03aef stopped and removed successfully
      yugabyte.go:112: stopping and removing node: yuga-tablet-207abe7e-a0e0-4ce0-a6b0-32eb46d6e96f
  *** Aborted at 1763470596 (unix time) try "date -d @1763470596" if you are using GNU date ***
  PC: @                0x0 (unknown)
  *** SIGTERM (@0x1) received by PID 7 (TID 0x7fcd83218340) from PID 1; stack trace: ***
      @     0x7fcd8226d371 __pthread_cond_wait
      @     0x55e9ae47798b yb::TerminationMonitor::WaitForTermination()
      @     0x55e9afaec969 yb::tserver::TabletServerMain()
      @     0x55e9ae035ce4 main
      @     0x7fcd824b97e5 __libc_start_main
      @     0x55e9ade3f02e _start
      yugabyte.go:112: Container 9394555df16a9399a707e1e2902ad53237d036b21541b9f18779f8a718837ab7 stopped and removed successfully
      yugabyte.go:97: network sc_yuga_net_9d563c00-ca67-434e-b024-019ca99ff425 removed successfully
```

--- a/docs/setup.md
+++ b/docs/setup.md
@@ -126,7 +126,7 @@ Example: Start Yugabyte via docker
 docker run --name sc_yugabyte_unit_tests \
   --platform linux/amd64 \
   -p 5433:5433 \
-  -d yugabytedb/yugabyte:2.20.2.3-b2 \
+  -d yugabytedb/yugabyte:2025.2.0.1-b1 \
   bin/yugabyted start \
   --background=false \
   --advertise_address=0.0.0.0 \

--- a/integration/runner/cluster_controller.go
+++ b/integration/runner/cluster_controller.go
@@ -49,7 +49,7 @@ func (cc *DBClusterController) getNodesConnectionsByRole(
 	t.Helper()
 	endpoints := make([]*connection.Endpoint, 0, len(cc.nodes))
 	for _, node := range cc.IterNodesByRole(role) {
-		endpoints = append(endpoints, node.GetContainerConnectionDetails(t))
+		endpoints = append(endpoints, node.GetHostMappedEndpoint(t))
 	}
 	return dbtest.NewConnection(endpoints...)
 }
@@ -66,7 +66,7 @@ func (cc *DBClusterController) getNodesConnections(t *testing.T) *dbtest.Connect
 	t.Helper()
 	endpoints := make([]*connection.Endpoint, len(cc.nodes))
 	for i, node := range cc.nodes {
-		endpoints[i] = node.GetContainerConnectionDetails(t)
+		endpoints[i] = node.GetHostMappedEndpoint(t)
 	}
 
 	return dbtest.NewConnection(endpoints...)

--- a/integration/runner/postgres.go
+++ b/integration/runner/postgres.go
@@ -9,7 +9,6 @@ package runner
 import (
 	"context"
 	"fmt"
-	"runtime"
 	"testing"
 
 	"github.com/google/uuid"
@@ -65,10 +64,6 @@ type (
 // and returns its connection properties.
 func StartPostgresCluster(ctx context.Context, t *testing.T) (*PostgresClusterController, *dbtest.Connection) {
 	t.Helper()
-
-	if runtime.GOOS != linuxOS {
-		t.Skip("Container IP access not supported on non-linux Docker")
-	}
 
 	cluster := &PostgresClusterController{}
 
@@ -140,6 +135,9 @@ func (cc *PostgresClusterController) createNode(
 			"ALLOW_EMPTY_PASSWORD=yes",
 		}, nodeCreationOpts.additionalEnvs...),
 	}
+	// Expose the PostgreSQL port via host port mapping so the test client
+	// can connect through a host-accessible address on all platforms.
+	node.ExposePort("5432")
 	cc.nodes = append(cc.nodes, node)
 
 	return node

--- a/integration/runner/yugabyte.go
+++ b/integration/runner/yugabyte.go
@@ -13,7 +13,6 @@ import (
 	"net"
 	"path"
 	"regexp"
-	"runtime"
 	"strings"
 	"testing"
 	"time"
@@ -50,7 +49,7 @@ type (
 
 const (
 	// latest LTS.
-	defaultImage = "yugabytedb/yugabyte:2024.2.4.0-b89"
+	defaultImage = "yugabytedb/yugabyte:2025.2.0.1-b1"
 
 	networkPrefix = "sc_yuga_net_"
 	masterPort    = "7100"
@@ -77,12 +76,12 @@ func StartYugaCluster(ctx context.Context, t *testing.T, numberOfMasters, number
 ) {
 	t.Helper()
 
-	if runtime.GOOS != linuxOS {
-		t.Skip("Container IP access not supported on non-linux Docker")
-	}
-
 	t.Logf("starting yuga cluster with (%d) masters and (%d) tablets ", numberOfMasters, numberOfTablets)
 
+	// YugabyteDB uses Raft consensus for data replication across tablet
+	// servers. With RF=3, each Raft group tolerates 1 node failure (majority
+	// quorum = 2 of 3). With fewer than 3 tablets, RF=3 is impossible, so we
+	// fall back to RF=1 (no fault tolerance, single-node mode).
 	rf := 1
 	if numberOfTablets >= 3 {
 		rf = 3
@@ -92,6 +91,9 @@ func StartYugaCluster(ctx context.Context, t *testing.T, numberOfMasters, number
 		replicationFactor: rf,
 		networkName:       fmt.Sprintf("%s%s", networkPrefix, uuid.NewString()),
 	}
+	// A dedicated Docker network enables container-to-container communication
+	// via container names (Docker DNS). Masters and tservers use each other's
+	// container names as RPC addresses for Raft consensus and heartbeats.
 	test.CreateDockerNetwork(t, cluster.networkName)
 	t.Cleanup(func() {
 		test.RemoveDockerNetwork(t, cluster.networkName)
@@ -112,9 +114,12 @@ func StartYugaCluster(ctx context.Context, t *testing.T, numberOfMasters, number
 		cluster.stopAndRemoveCluster(t)
 	})
 
-	// The master nodes are not involved in DB communication;
-	// the application connects to the tablet servers.
+	// In YugabyteDB, masters manage cluster metadata (tablet locations, schema,
+	// Raft leadership) while tservers handle all SQL read/write operations.
+	// The application only connects to tservers for data access.
 	clusterConnection := cluster.getNodesConnectionsByRole(t, TabletNode)
+	// LoadBalance distributes new connections across all tservers in the
+	// cluster instead of sending all traffic to a single node.
 	clusterConnection.LoadBalance = true
 
 	return cluster, clusterConnection
@@ -126,7 +131,15 @@ func (cc *YugaClusterController) createNode(role string) {
 		Image:        defaultImage,
 		Role:         role,
 		DatabaseType: dbtest.YugaDBType,
-		Network:      cc.networkName,
+		// All nodes join the same Docker network so they can resolve each
+		// other by container name for inter-node RPC (Raft, heartbeats).
+		Network: cc.networkName,
+	}
+	// Expose the YSQL port via host port mapping so the test client can reach
+	// the tablet servers through a host-accessible address on all platforms.
+	// Only tablets need this because masters are not involved in DB communication.
+	if role == TabletNode {
+		node.ExposePort("5433")
 	}
 	cc.nodes = append(cc.nodes, node)
 }
@@ -134,6 +147,10 @@ func (cc *YugaClusterController) createNode(role string) {
 func (cc *YugaClusterController) startNodes(ctx context.Context, t *testing.T) {
 	t.Helper()
 	masterAddresses := cc.getMasterAddresses()
+
+	// Configure all nodes before starting any containers. Each node needs
+	// the full list of master addresses so it can join the Raft consensus
+	// group (masters) or discover where to register (tservers).
 	for _, n := range cc.IterNodesByRole(MasterNode) {
 		n.Cmd = nodeConfig(t, nodeConfigParameters{
 			n.Role,
@@ -141,15 +158,7 @@ func (cc *YugaClusterController) startNodes(ctx context.Context, t *testing.T) {
 			masterAddresses,
 			cc.replicationFactor,
 		})
-		n.StartContainer(ctx, t)
 	}
-
-	expectedAlive := len(maps.Collect(cc.IterNodesByRole(MasterNode)))
-	require.EventuallyWithT(t, func(ct *assert.CollectT) {
-		actualAlive := cc.getNumberOfAliveMasters(t)
-		require.Equal(ct, expectedAlive, actualAlive)
-	}, time.Minute, time.Millisecond*100)
-
 	for _, n := range cc.IterNodesByRole(TabletNode) {
 		n.Cmd = nodeConfig(t, nodeConfigParameters{
 			n.Role,
@@ -157,12 +166,63 @@ func (cc *YugaClusterController) startNodes(ctx context.Context, t *testing.T) {
 			masterAddresses,
 			cc.replicationFactor,
 		})
+	}
+
+	// Start tservers BEFORE masters.
+	//
+	// Observation: when masters started first, the DB resiliency tests that
+	// remove a single tablet server would permanently stall all transactions.
+	// Inspecting the cluster with `yb-admin list_tablets system transactions`
+	// showed the global transaction table (system.transactions) was created
+	// with RF=2 instead of the expected RF=3.
+	//
+	// Root cause: the master leader creates system.transactions during its
+	// initialization, using RF = min(num_registered_tservers, cluster_RF).
+	// When masters start first, the leader initializes before all tservers
+	// have sent their first heartbeat, so it sees only 2 registered tservers
+	// and creates the table with RF=2. Note: the flag
+	// `txn_table_wait_min_ts_count` does NOT help here — it only controls
+	// a background task code path, not the leader initialization code path.
+	//
+	// Consequence: with RF=2, each Raft group for the transaction table has
+	// only 2 replicas. Removing 1 tablet server drops some groups to 1/2
+	// replicas, which is below majority quorum (need 2/2 alive). This
+	// permanently stalls all distributed transactions. The table's RF
+	// cannot be fixed after creation — modify_table_placement_info is
+	// explicitly blocked for system.transactions by YugabyteDB.
+	//
+	// Fix: by starting tservers first, they are already retrying master
+	// connections when the master comes up. All tservers register within
+	// the first heartbeat window, so the leader sees 3 registered tservers
+	// and creates the table with the full RF=3.
+	for _, n := range cc.IterNodesByRole(TabletNode) {
+		n.StartContainer(ctx, t)
+	}
+	for _, n := range cc.IterNodesByRole(MasterNode) {
 		n.StartContainer(ctx, t)
 	}
 
+	// Wait for all masters to form a Raft quorum and elect a leader.
+	// Until a leader is elected, the cluster cannot accept tserver
+	// registrations, create system tables, or serve any metadata requests.
+	expectedMasters := len(maps.Collect(cc.IterNodesByRole(MasterNode)))
+	require.EventuallyWithT(t, func(ct *assert.CollectT) {
+		require.Equal(ct, expectedMasters, strings.Count(cc.listAllMasters(t), "ALIVE"))
+	}, time.Minute, time.Millisecond*100)
+
+	// Wait for each tserver to finish bootstrapping (syncing data to disk).
+	// Until this completes, the tserver is not ready to serve SQL queries.
 	for _, n := range cc.IterNodesByRole(TabletNode) {
 		n.EnsureNodeReadinessByLogs(t, dbtest.YugabyteTabletNodeReadinessOutput)
 	}
+
+	// Wait for all tservers to register with the master via heartbeats.
+	// The master must know about all tservers before the cluster is ready
+	// to serve queries reliably.
+	expectedTablets := len(maps.Collect(cc.IterNodesByRole(TabletNode)))
+	require.EventuallyWithT(t, func(ct *assert.CollectT) {
+		require.Equal(ct, expectedTablets, strings.Count(cc.runYBAdmin(t, "list_all_tablet_servers"), "ALIVE"))
+	}, time.Minute, time.Millisecond*500)
 }
 
 func (cc *YugaClusterController) getMasterAddresses() string {
@@ -180,25 +240,25 @@ func (cc *YugaClusterController) getLeaderMasterName(t *testing.T) string {
 	return found[1]
 }
 
-func (cc *YugaClusterController) getNumberOfAliveMasters(t *testing.T) int {
-	t.Helper()
-	return strings.Count(cc.listAllMasters(t), "ALIVE")
-}
-
 func (cc *YugaClusterController) listAllMasters(t *testing.T) string {
 	t.Helper()
-	cmd := []string{
+	output := cc.runYBAdmin(t, "list_all_masters")
+	require.NotEmpty(t, output, "Could not get yb-admin output from any node")
+	return output
+}
+
+func (cc *YugaClusterController) runYBAdmin(t *testing.T, args ...string) string {
+	t.Helper()
+	cmd := append([]string{
 		"/home/yugabyte/bin/yb-admin",
 		"-master_addresses", cc.getMasterAddresses(),
-		"list_all_masters",
-	}
+	}, args...)
 	var output string
 	for _, n := range cc.nodes {
 		if output = n.ExecuteCommand(t, cmd); output != "" {
 			break
 		}
 	}
-	require.NotEmpty(t, output, "Could not get yb-admin output from any node")
 	return output
 }
 
@@ -228,13 +288,31 @@ func nodeConfig(t *testing.T, params nodeConfigParameters) []string {
 	switch params.role {
 	case MasterNode:
 		return append(nodeCommonConfig("yb-master", params.nodeName, masterPort),
+			// List of all master nodes so they can form a Raft group.
 			"--master_addresses", params.masterAddresses,
+			// Desired number of replicas for system and user tables.
 			fmt.Sprintf("--replication_factor=%d", params.replicationFactor),
 		)
 	case TabletNode:
 		return append(nodeCommonConfig("yb-tserver", params.nodeName, tabletPort),
+			// Enable the YSQL (PostgreSQL-compatible) query layer.
 			"--start_pgsql_proxy",
+			// Master addresses so the tserver can register and discover tablets.
 			"--tserver_master_addrs", params.masterAddresses,
+			// Limit shards per tserver to reduce resource usage in tests.
+			"--yb_num_shards_per_tserver=1",
+			// By default YSQL binds to the container hostname only. With host
+			// port forwarding the traffic arrives on 0.0.0.0 inside the
+			// container, so we must listen on all interfaces.
+			"--pgsql_proxy_bind_address=0.0.0.0",
+			// glog (used by yb-tserver) defaults to writing logs to files.
+			// EnsureNodeReadinessByLogs reads container stdout, so we redirect
+			// logs to stderr which Docker captures.
+			"--logtostderr",
+			// Enable read committed isolation level. Without this,
+			// YugabyteDB falls back to repeatable read for read committed
+			// transactions, which changes concurrency behavior.
+			"--yb_enable_read_committed_isolation=true",
 		)
 	default:
 		t.Fatalf("unknown role provided: %s", params.role)

--- a/integration/test/db_node_failure_handling_test.go
+++ b/integration/test/db_node_failure_handling_test.go
@@ -8,6 +8,7 @@ package test
 
 import (
 	"context"
+	"runtime"
 	"testing"
 	"time"
 
@@ -20,6 +21,7 @@ import (
 )
 
 const (
+	linuxOS = "linux"
 	// Tablet is a tablet (data node) in the cluster.
 	Tablet = 1 << iota
 	// NonLeaderMaster is a master node that does not hold the leader role.
@@ -29,7 +31,13 @@ const (
 )
 
 func TestDBResiliencyYugabyteScenarios(t *testing.T) {
-	t.Parallel()
+	// Parallel is only enabled on Linux. On macOS each subtest spins up a
+	// 6-container YugabyteDB cluster (3 masters + 3 tablets). Running multiple
+	// clusters concurrently exhausts Docker Desktop resources and causes
+	// container startup timeouts.
+	if runtime.GOOS == linuxOS {
+		t.Parallel()
+	}
 
 	for _, sc := range []struct {
 		name   string
@@ -58,7 +66,9 @@ func TestDBResiliencyYugabyteScenarios(t *testing.T) {
 	} {
 		scenario := sc
 		t.Run(scenario.name, func(t *testing.T) {
-			t.Parallel()
+			if runtime.GOOS == linuxOS {
+				t.Parallel()
+			}
 			clusterController, clusterConnection := runner.StartYugaCluster(createInitContext(t), t, 3, 3)
 			c := registerAndCreateRuntime(t, clusterConnection)
 			waitForCommittedTxs(t, c, 10_000)
@@ -77,7 +87,10 @@ func TestDBResiliencyYugabyteScenarios(t *testing.T) {
 }
 
 func TestDBResiliencyPrimaryPostgresNodeCrash(t *testing.T) {
-	t.Parallel()
+	// See TestDBResiliencyYugabyteScenarios for why parallel is Linux-only.
+	if runtime.GOOS == linuxOS {
+		t.Parallel()
+	}
 
 	clusterController, clusterConnection := runner.StartPostgresCluster(createInitContext(t), t)
 
@@ -90,7 +103,10 @@ func TestDBResiliencyPrimaryPostgresNodeCrash(t *testing.T) {
 }
 
 func TestDBResiliencySecondaryPostgresNodeCrash(t *testing.T) {
-	t.Parallel()
+	// See TestDBResiliencyYugabyteScenarios for why parallel is Linux-only.
+	if runtime.GOOS == linuxOS {
+		t.Parallel()
+	}
 
 	clusterController, clusterConnection := runner.StartPostgresCluster(createInitContext(t), t)
 
@@ -131,9 +147,12 @@ func waitForCommittedTxs(t *testing.T, c *runner.CommitterRuntime, waitForCount 
 	require.Zero(t, c.CountAlternateStatus(t, committerpb.Status_COMMITTED))
 }
 
+// createInitContext returns a context with a timeout covering the full cluster
+// lifecycle: container image pulls, sequential startup of 6 nodes (on macOS),
+// transaction processing, node removal, and recovery verification.
 func createInitContext(t *testing.T) context.Context {
 	t.Helper()
-	ctx, cancel := context.WithTimeout(t.Context(), 3*time.Minute)
+	ctx, cancel := context.WithTimeout(t.Context(), 8*time.Minute)
 	t.Cleanup(cancel)
 
 	return ctx

--- a/scripts/get-and-start-yuga-mac.sh
+++ b/scripts/get-and-start-yuga-mac.sh
@@ -14,7 +14,7 @@ YUGA_DIR=${YUGA_DIR:="$HOME/Library/bin/yugabyte"}
 if [ ! -d "$YUGA_DIR" ]; then
   echo "Downloading and extracting YugabyteDB"
   mkdir -p "$YUGA_DIR"
-  curl https://downloads.yugabyte.com/releases/2.20.7.0/yugabyte-2.20.7.0-b58-darwin-x86_64.tar.gz |
+  curl https://downloads.yugabyte.com/releases/2025.2.0.1/yugabyte-2025.2.0.1-b1-darwin-x86_64.tar.gz |
     tar --strip=1 -xz -C "$YUGA_DIR"
 fi
 

--- a/scripts/get-and-start-yuga.sh
+++ b/scripts/get-and-start-yuga.sh
@@ -10,7 +10,7 @@ YUGA_DIR=${YUGA_DIR:="$HOME/bin/yugabyte"}
 if [ ! -d "$YUGA_DIR" ]; then
   echo "Downloading and extracting YugabyteDB"
   mkdir -p "$YUGA_DIR"
-  curl https://downloads.yugabyte.com/releases/2.20.7.0/yugabyte-2.20.7.0-b58-linux-x86_64.tar.gz |
+  curl https://downloads.yugabyte.com/releases/2025.2.0.1/yugabyte-2025.2.0.1-b1-linux-x86_64.tar.gz |
     tar --strip=1 -xz -C "$YUGA_DIR"
 
   cd "$YUGA_DIR" || exit 1

--- a/service/vc/dbinit.go
+++ b/service/vc/dbinit.go
@@ -15,6 +15,8 @@ import (
 	"github.com/cockroachdb/errors"
 	"github.com/hyperledger/fabric-x-common/api/committerpb"
 	"github.com/yugabyte/pgx/v4/pgxpool"
+
+	"github.com/hyperledger/fabric-x-committer/utils/dbconn"
 )
 
 const (
@@ -55,6 +57,8 @@ func NewDatabasePool(ctx context.Context, config *DatabaseConfig) (*pgxpool.Pool
 
 	poolConfig.MaxConns = config.MaxConnections
 	poolConfig.MinConns = config.MinConnections
+
+	dbconn.ConfigureConnReadDeadline(poolConfig)
 
 	var pool *pgxpool.Pool
 	if retryErr := config.Retry.Execute(ctx, func() error {

--- a/service/vc/dbtest/connection.go
+++ b/service/vc/dbtest/connection.go
@@ -88,6 +88,8 @@ func (c *Connection) open(ctx context.Context) (*pgxpool.Pool, error) {
 	poolConfig.MaxConns = 1
 	poolConfig.MinConns = 1
 
+	dbconn.ConfigureConnReadDeadline(poolConfig)
+
 	var pool *pgxpool.Pool
 	if retryErr := defaultRetry.Execute(ctx, func() error {
 		pool, err = pgxpool.ConnectConfig(ctx, poolConfig)

--- a/service/vc/dbtest/container.go
+++ b/service/vc/dbtest/container.go
@@ -28,7 +28,7 @@ import (
 )
 
 const (
-	defaultYugabyteImage            = "yugabytedb/yugabyte:2.20.7.0-b58"
+	defaultYugabyteImage            = "yugabytedb/yugabyte:2025.2.0.1-b1"
 	defaultPostgresImage            = "postgres:16.9-alpine3.21"
 	defaultDBDeploymentTemplateName = test.DockerNamesPrefix + "_%s_unit_tests"
 
@@ -207,13 +207,20 @@ func (dc *DatabaseContainer) createContainer(ctx context.Context, t *testing.T) 
 		return
 	}
 
-	// Pull the image if not exist
-	require.NoError(t, dc.client.PullImage(docker.PullImageOptions{
-		Context:      ctx,
-		Repository:   dc.Image,
-		Tag:          dc.Tag,
-		OutputStream: os.Stdout,
-	}, docker.AuthConfiguration{}))
+	// Pull the image only if it doesn't exist locally. This avoids
+	// unnecessary Docker Hub requests that count against the rate limit.
+	imageRef := dc.Image
+	if dc.Tag != "" {
+		imageRef += ":" + dc.Tag
+	}
+	if _, err = dc.client.InspectImage(imageRef); err != nil {
+		require.NoError(t, dc.client.PullImage(docker.PullImageOptions{
+			Context:      ctx,
+			Repository:   dc.Image,
+			Tag:          dc.Tag,
+			OutputStream: os.Stdout,
+		}, docker.AuthConfiguration{}))
+	}
 
 	// Create the container instance
 	container, err := dc.client.CreateContainer(
@@ -307,11 +314,41 @@ func (dc *DatabaseContainer) GetContainerConnectionDetails(
 	return connection.CreateEndpointHP(ipAddress, dc.DbPort.Port())
 }
 
+// ExposePort adds a host port mapping for the given container port (e.g. "5433")
+// with an auto-assigned host port. This allows the test client to reach the
+// container through a host-accessible address on all platforms.
+func (dc *DatabaseContainer) ExposePort(port string) {
+	p := docker.Port(port + "/tcp")
+	if dc.PortBinds == nil {
+		dc.PortBinds = make(map[docker.Port][]docker.PortBinding)
+	}
+	dc.PortBinds[p] = []docker.PortBinding{{HostIP: "0.0.0.0", HostPort: ""}}
+}
+
+// GetHostMappedEndpoint inspects the container and returns a host-accessible endpoint
+// using the auto-assigned host port mapping.
+func (dc *DatabaseContainer) GetHostMappedEndpoint(t *testing.T) *connection.Endpoint {
+	t.Helper()
+	container, err := dc.client.InspectContainerWithOptions(docker.InspectContainerOptions{
+		Context: t.Context(),
+		ID:      dc.containerID,
+	})
+	require.NoError(t, err)
+
+	bindings := container.NetworkSettings.Ports[dc.DbPort]
+	require.NotEmpty(t, bindings, "no host port mapping found for port %s", dc.DbPort)
+
+	hostIP := bindings[0].HostIP
+	if hostIP == "0.0.0.0" {
+		hostIP = "127.0.0.1"
+	}
+	return connection.CreateEndpointHP(hostIP, bindings[0].HostPort)
+}
+
 // streamLogs streams the container output to the requested stream.
 func (dc *DatabaseContainer) streamLogs(t *testing.T) {
 	t.Helper()
 	logOptions := docker.LogsOptions{
-		//nolint:usetesting //t.Context finished after the function call, which is causing an unexpected crash.
 		Context:      context.Background(),
 		Container:    dc.containerID,
 		Follow:       true,

--- a/utils/dbconn/database_connection.go
+++ b/utils/dbconn/database_connection.go
@@ -7,9 +7,13 @@ SPDX-License-Identifier: Apache-2.0
 package dbconn
 
 import (
+	"context"
 	"fmt"
+	"net"
+	"time"
 
 	"github.com/cockroachdb/errors"
+	"github.com/yugabyte/pgx/v4/pgxpool"
 
 	"github.com/hyperledger/fabric-x-committer/utils/connection"
 )
@@ -29,6 +33,13 @@ type (
 		EndpointsString string
 		LoadBalance     bool
 		TLS             DatabaseTLSConfig
+	}
+
+	// deadlineConn wraps a net.Conn with a per-read deadline so that reads
+	// on dead connections fail instead of hanging indefinitely (on MacOS's Docker VM).
+	deadlineConn struct {
+		net.Conn
+		readTimeout time.Duration
 	}
 )
 
@@ -57,4 +68,48 @@ func DataSourceName(d DataSourceNameParams) (string, error) {
 		ret += "&load_balance=true"
 	}
 	return ret, nil
+}
+
+// ConfigureConnReadDeadline configures pool connections with a per-read
+// deadline so that reads on dead connections fail instead of hanging
+// indefinitely.
+//
+// We use a custom DialFunc that wraps every net.Conn with deadlineConn,
+// which calls SetReadDeadline before each Read. Pool callbacks like
+// AfterConnect or BeforeAcquire only fire once, so they can only set a
+// single absolute deadline that expires even if the connection is healthy.
+// Wrapping net.Conn lets us reset the deadline on every Read call.
+//
+// The read timeout is set to 60 seconds, which is more than enough for
+// read-set validation and commit operations. This can be made configurable
+// if needed.
+//
+// On macOS, Docker Desktop runs a userspace TCP proxy inside a Linux VM.
+// When a container dies the proxy does not immediately close forwarded
+// connections, so reads hang indefinitely. The deadlineConn wrapper ensures
+// stuck reads fail within the timeout, allowing the pool to replace them.
+// On Linux, reads already fail immediately via TCP RST when a node dies,
+// so the wrapper is harmless.
+func ConfigureConnReadDeadline(poolConfig *pgxpool.Config) {
+	dialer := &net.Dialer{
+		KeepAlive: 3 * time.Second,
+		Timeout:   10 * time.Second,
+	}
+	poolConfig.ConnConfig.DialFunc = func(ctx context.Context, network, addr string) (net.Conn, error) {
+		conn, err := dialer.DialContext(ctx, network, addr)
+		if err != nil {
+			return nil, err
+		}
+		return &deadlineConn{Conn: conn, readTimeout: 60 * time.Second}, nil
+	}
+	// Lower from the default (1 min) so the pool detects dead idle
+	// connections faster.
+	poolConfig.HealthCheckPeriod = 30 * time.Second
+}
+
+func (c *deadlineConn) Read(b []byte) (int, error) {
+	if err := c.SetReadDeadline(time.Now().Add(c.readTimeout)); err != nil {
+		return 0, err
+	}
+	return c.Conn.Read(b)
 }

--- a/utils/dbconn/database_connection_test.go
+++ b/utils/dbconn/database_connection_test.go
@@ -1,0 +1,173 @@
+/*
+Copyright IBM Corp. All Rights Reserved.
+
+SPDX-License-Identifier: Apache-2.0
+*/
+
+package dbconn
+
+import (
+	"net"
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+	"github.com/yugabyte/pgx/v4/pgxpool"
+
+	"github.com/hyperledger/fabric-x-committer/utils/channel"
+	"github.com/hyperledger/fabric-x-committer/utils/connection"
+)
+
+func TestDataSourceName(t *testing.T) {
+	t.Parallel()
+
+	params := DataSourceNameParams{
+		Username:        "user",
+		Password:        "pass",
+		Database:        "mydb",
+		EndpointsString: "localhost:5433",
+	}
+
+	tests := []struct {
+		name        string
+		modify      func(*DataSourceNameParams)
+		expectedDSN string
+		expectedErr string
+	}{
+		{
+			name:        "none TLS mode",
+			modify:      func(p *DataSourceNameParams) { p.TLS.Mode = connection.NoneTLSMode },
+			expectedDSN: "postgres://user:pass@localhost:5433/mydb?sslmode=disable",
+		},
+		{
+			name:        "unmentioned TLS mode",
+			modify:      func(p *DataSourceNameParams) { p.TLS.Mode = connection.UnmentionedTLSMode },
+			expectedDSN: "postgres://user:pass@localhost:5433/mydb?sslmode=disable",
+		},
+		{
+			name: "one-side TLS mode",
+			modify: func(p *DataSourceNameParams) {
+				p.TLS.Mode = connection.OneSideTLSMode
+				p.TLS.CACertPath = "/path/to/ca.pem"
+			},
+			expectedDSN: "postgres://user:pass@localhost:5433/mydb?sslmode=verify-full&sslrootcert=/path/to/ca.pem",
+		},
+		{
+			name:        "mutual TLS mode returns error",
+			modify:      func(p *DataSourceNameParams) { p.TLS.Mode = connection.MutualTLSMode },
+			expectedErr: "unsupportted db tls mode",
+		},
+		{
+			name:        "unknown TLS mode returns error",
+			modify:      func(p *DataSourceNameParams) { p.TLS.Mode = "bogus" },
+			expectedErr: "unknown TLS mode",
+		},
+		{
+			name: "load balance enabled",
+			modify: func(p *DataSourceNameParams) {
+				p.TLS.Mode = connection.NoneTLSMode
+				p.LoadBalance = true
+			},
+			expectedDSN: "postgres://user:pass@localhost:5433/mydb?sslmode=disable&load_balance=true",
+		},
+		{
+			name:        "load balance disabled",
+			modify:      func(p *DataSourceNameParams) { p.TLS.Mode = connection.NoneTLSMode },
+			expectedDSN: "postgres://user:pass@localhost:5433/mydb?sslmode=disable",
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			t.Parallel()
+
+			p := params
+			tt.modify(&p)
+
+			dsn, err := DataSourceName(p)
+
+			if tt.expectedErr != "" {
+				require.ErrorContains(t, err, tt.expectedErr)
+				return
+			}
+			require.NoError(t, err)
+			assert.Equal(t, tt.expectedDSN, dsn)
+		})
+	}
+}
+
+func TestConfigureConnReadDeadline(t *testing.T) {
+	t.Parallel()
+
+	poolConfig, err := pgxpool.ParseConfig("postgres://user:pass@localhost:5433/mydb")
+	require.NoError(t, err)
+
+	ConfigureConnReadDeadline(poolConfig)
+
+	assert.Equal(t, 30*time.Second, poolConfig.HealthCheckPeriod)
+	assert.NotNil(t, poolConfig.ConnConfig.DialFunc)
+}
+
+func TestDeadlineConn(t *testing.T) {
+	t.Parallel()
+
+	t.Run("successful read resets deadline each time", func(t *testing.T) {
+		t.Parallel()
+
+		server, client := tcpPair(t)
+
+		dc := &deadlineConn{Conn: client, readTimeout: 5 * time.Second}
+
+		expected := []byte("hello")
+		_, err := server.Write(expected)
+		require.NoError(t, err)
+
+		buf := make([]byte, len(expected))
+		n, err := dc.Read(buf)
+		require.NoError(t, err)
+		assert.Equal(t, expected, buf[:n])
+	})
+
+	t.Run("read times out on stalled connection", func(t *testing.T) {
+		t.Parallel()
+
+		_, client := tcpPair(t)
+
+		dc := &deadlineConn{Conn: client, readTimeout: 50 * time.Millisecond}
+
+		buf := make([]byte, 64)
+		_, err := dc.Read(buf)
+		require.Error(t, err)
+
+		var netErr net.Error
+		require.ErrorAs(t, err, &netErr)
+		assert.True(t, netErr.Timeout())
+	})
+}
+
+// tcpPair creates a connected pair of TCP connections for testing.
+func tcpPair(t *testing.T) (server, client net.Conn) {
+	t.Helper()
+
+	ln, err := net.Listen("tcp", "127.0.0.1:0")
+	require.NoError(t, err)
+	t.Cleanup(func() { _ = ln.Close() })
+
+	accepted := channel.NewReaderWriter(t.Context(), make(chan net.Conn, 1))
+	go func() {
+		conn, errA := ln.Accept()
+		if errA == nil {
+			accepted.Write(conn)
+		}
+	}()
+
+	client, err = net.Dial("tcp", ln.Addr().String())
+	require.NoError(t, err)
+	t.Cleanup(func() { _ = client.Close() })
+
+	server, _ = accepted.Read()
+	t.Cleanup(func() { _ = server.Close() })
+
+	return server, client
+}

